### PR TITLE
Update label files while project is loaded.

### PR DESCRIPTION
**Pull request by SimoTw**

### CODE REVIEW

File: src/redux/actions/projectActions.ts [hunk 1]
```diff
@@ -211,18 +213,54 @@ export function deleteAsset(project: IProject, assetMetadata: IAssetMetadata): (
...
+
+export async function checkAndUpdateSchema (project: IProject): Promise<IProject> {
+    let shouldAssetsUpdate = false;
+    let updatedProject;
+    const {assets} = project;
+    if (_.isPlainObject(assets)) {
+        const assetService = new AssetService(project);
+        const assetMetadatas: IAssetMetadata[] = await Promise.all(Object.values(assets).map(async (asset) => await assetService.getAssetMetadata(asset)));
+        await Promise.all(assetMetadatas.map(async (assetMetadata) => {
+            if (_.isPlainObject(assetMetadata.labelData) && assetMetadata.labelData?.$schema !== constants.labelsSchema) {
```
Comment (buddhawang) [inline src/redux/actions/projectActions.ts]: suggestion: we should check if it's supported schema URLs here.

File: src/redux/actions/projectActions.ts [hunk 2]
```diff
@@ -211,18 +213,54 @@ export function deleteAsset(project: IProject, assetMetadata: IAssetMetadata): (
...
+                assetMetadata.labelData = {...assetMetadata.labelData, "$schema": constants.labelsSchema};
+                await assetService.save(assetMetadata);
+            }
+        }))
+        const updatedAssets = {...assets};
+        for (const [assetID, asset] of Object.entries(assets)) {
+            if (asset.schema !== constants.labelsSchema) {
```
Comment (buddhawang) [inline src/redux/actions/projectActions.ts]: suggestion: we should check if it's supported schema URLs here.

File: src/redux/actions/projectActions.ts [hunk 3]
```diff
@@ -211,18 +213,54 @@ export function deleteAsset(project: IProject, assetMetadata: IAssetMetadata): (
...
+                updatedAssets[assetID] = {...assets[assetID], schema: constants.labelsSchema};
```
Comment (buddhawang) [inline src/redux/actions/projectActions.ts]: question: set `shouldAssetsUpdate = true` here?

File: src/redux/actions/projectActions.ts [hunk 4]
```diff
@@ -211,18 +211,31 @@ export function deleteAsset(project: IProject, assetMetadata: IAssetMetadata): (
     };
 }
 
+
+
+
```
Comment (buddhawang) [inline src/redux/actions/projectActions.ts:216]: nit: remove extra blank lines.

File: src/services/projectService.ts [hunk 5]
```diff
@@ -411,19 +411,19 @@ export default class ProjectService implements IProjectService {
...
      * @param storageProvider the storage we're trying to save the project
      */
-    private async saveFieldsFile(project: IProject, storageProvider: IStorageProvider) {
+    public async saveFieldsFile(project: IProject, storageProvider: IStorageProvider) {
         Guard.null(project);
         Guard.null(project.tags);
 
         const definitions = {};
         const fieldInfo = {};
-        fieldInfo["$schema"] = "http://www.azure.com/schema/formrecognizer/fields.json"
```
Comment (buddhawang) [inline src/services/projectService.ts:426]: question: I thought this was fixed? had you rebase your branch?

File: src/react/components/pages/train/trainPage.tsx [hunk 6]
```diff
@@ -314,6 +314,7 @@ export default class TrainPage extends React.Component<ITrainPageProps, ITrainPa
         });
 
         this.trainProcess().then(async (trainResult) => {
+            await AssetService.checkAndUpdateSchema(this.props.project);
```
Comment (buddhawang) [inline src/react/components/pages/train/trainPage.tsx:317]: question: I'm a bit confused the reason to checkAndUpdateSchema here after trainResult is returned, can you elaborate more?

File: src/redux/actions/projectActions.ts [hunk 7]
```diff
@@ -211,18 +211,31 @@ export function deleteAsset(project: IProject, assetMetadata: IAssetMetadata): (
...
  */
-export function loadAssets(project: IProject): (dispatch: Dispatch) => Promise<IAsset[]> {
-    return async (dispatch: Dispatch) => {
+export function loadAssets(project: IProject): (dispatch: Dispatch, getState: () => IApplicationState) => Promise<IAsset[]> {
+    return async (dispatch: Dispatch, getState: () => IApplicationState) => {
         const assetService = new AssetService(project);
         const assets = await assetService.getAssets();
+        let shouldAssetsUpdate = false;
+        for (const asset of assets) {
+            if (asset.schema !== constants.labelsSchema) {
```
Comment (buddhawang) [inline src/redux/actions/projectActions.ts:227]: suggestion: use `shouldSchemaUpdate `?

**Review by buddhawang — COMMENTED** [review]
Add some comments.

File: src/react/components/pages/train/trainPage.tsx (hunk 6 above)
Comment (SimoTw) [reply src/react/components/pages/train/trainPage.tsx]: Checked. It's redundant.

**Pull request merged**